apps/designer/app/shared/copy-paste/use-instance-copy-paste.ts

```diff
@@ -0,0 +1,130 @@
+import { utils } from "@webstudio-is/project";
+import type { Instance, UserProp } from "@webstudio-is/react-sdk";
+import { useEffect, useRef } from "react";
+import { type InstanceCopyData, serialize, deserialize } from "./serialize";
+
+type Props = {
+  selectedInstanceData?: InstanceCopyData | undefined;
+  allowAnyTarget?: boolean;
+  onPaste?: (instance: Instance, props?: UserProp[]) => void;
+  onCut?: (instance: Instance) => void;
+};
+
+export const useInstanceCopyPaste = (props: Props): void => {
```

Comment:
btw if we want to some day render any framework on canvas, we need to start writing down the logic as much as possible outside of react stuff and then only wrap what's needed with a react hook, I think majority of this logic is universal, only useEffect/cleanup is react specific

Reply:
Not sure it's a good idea to make code more complex by adding indirection just in case we need it in some far future

Reply:
its not that far I think, within the next 6month, but also, I feel like it can be nicely written, using good patterns and this would show us a way forward with this, otherwise everyone will think this way

Reply:
Extracted